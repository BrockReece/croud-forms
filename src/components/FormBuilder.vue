<script>
    export default {
        props: {
            value: {
                required: true,
            },
            schema: {
                required: true,
            },

            readOnly: {
                type: Boolean,
                default: false,
            },

            validations: {
                type: Object,
                default() {
                    return {}
                },
            },
        },

        render(h) {
            const schema = this.schema.map((field) => {
                if (field.children) {
                    return this.recur(h, field)
                }

                return this.buildNode(h, field)
            })

            return h('div', {
                class: 'ui form',
            }, schema)
        },

        methods: {
            buildNode(h, field) {
                return h('croud-form-field', {
                    props: {
                        field,
                        model: this.value[field.field_slug],
                        readOnly: this.readOnly || field.readOnly,
                        validation: this.validations[field.field_slug] || {},
                    },

                    on: {
                        'set-value': (val) => {
                            this.value[field.field_slug] = val
                        },
                    },
                })
            },

            recur(h, field) {
                const children = field.children.map((child) => {
                    if (child.children) {
                        return this.recur(h, child)
                    }

                    return this.buildNode(h, child)
                })

                return h('div', {
                    class: field.class,
                }, children)
            },
        },

        watch: {
            value: {
                deep: true,
                handler(val) {
                    this.$emit('input', val)
                },
            },
        },
    }
</script>
