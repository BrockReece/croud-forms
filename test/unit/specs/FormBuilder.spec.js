import { mount } from 'vue-test-utils'
import FormBuilder from '@/components/FormBuilder'

const schemas = {
    'should render simple form': [
        {
            field_name: 'Test',
            field_type: 'text',
            field_slug: 'a',
        },
    ],

    'should render children': [
        {
            class: 'two fields',
            children: [
                {
                    field_name: 'First name',
                    field_type: 'text',
                    field_slug: 'a',
                },
                {
                    field_name: 'Last name',
                    field_type: 'number',
                    field_slug: 'b',
                },
            ],
        },
    ],

    'should recursively render children': [
        {
            class: 'two fields',
            children: [
                {
                    field_name: 'First name',
                    field_type: 'text',
                    field_slug: 'a',
                },
                {
                    class: 'two fields',
                    children: [
                        {
                            field_name: 'Last name',
                            field_type: 'text',
                            field_slug: 'b',
                        },
                    ],
                },
            ],
        },
    ],
    'should render from method': [
        {
            field_name: 'Test',
            field_type: 'text',
            field_slug: 'a',
            render: (h) => h('div', {
                class: 'ui field',
            }, [
                h('label', 'Test'),
                h('input'),
            ]),
        },
    ],
}


describe('FormBuilder component', () => {
    Object.keys(schemas).forEach((test) => {
        it(test, () => {
            const vm = mount(FormBuilder, {
                propsData: {
                    schema: schemas[test],
                    value: {
                        a: 'foo',
                        b: 'bar',
                    },
                },
            })
            expect(vm.html()).toMatchSnapshot()
        })
    })
})
