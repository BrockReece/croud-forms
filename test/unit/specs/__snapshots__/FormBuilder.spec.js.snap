// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBuilder component should recursively render children 1`] = `
<div class="ui form">
    <div class="two fields">
        <croud-form-field></croud-form-field>
        <div class="two fields">
            <croud-form-field></croud-form-field>
        </div>
    </div>
</div>
`;

exports[`FormBuilder component should render children 1`] = `
<div class="ui form">
    <div class="two fields">
        <croud-form-field></croud-form-field>
        <croud-form-field></croud-form-field>
    </div>
</div>
`;

exports[`FormBuilder component should render from method 1`] = `
<div class="ui form">
    <div class="ui field"><label>Test</label><input></div>
</div>
`;

exports[`FormBuilder component should render simple form 1`] = `
<div class="ui form">
    <croud-form-field></croud-form-field>
</div>
`;
